src/hevm/src/EVM/Expr.hs

```diff
@@ -441,9 +441,14 @@ readStorage addr loc store@(ConcreteStore s) = case (addr, loc) of
     pure $ Lit val
   _ -> Just $ SLoad addr loc store
 readStorage addr' loc s@AbstractStore = Just $ SLoad addr' loc s
-readStorage addr' loc s@(SStore addr slot val prev) = case (addr, slot, addr', loc) of
-  (Lit _, Lit _, Lit _, Lit _) -> if loc == slot && addr == addr' then Just val else readStorage addr' loc prev
+readStorage addr' loc s@(SStore addr slot val prev) = case addr of
+  Lit _ -> if addr == addr'
+           then case slot of
+                  Lit _ -> if slot == loc then (Just val) else readStorage addr' loc prev
+                  _ -> Just $ SLoad addr' loc s
+           else readStorage addr' loc prev
```

Comment:
I feel that you can return `readStorage add' loc prev` only if both addresses are different literals. The above would return `readStorage addr' loc prev` skipping a potential relevant store if the addresses are syntactically different but semantically the same (for example one of them is `Lit 0x0` and the other `ReadWord idx (WriteWord idx (Lit 0x0) buf)` or some symbolic variable). The same goes for slot.

Reply:
nice catch, should be fixed now I think.